apps/ui-library/registry/default/examples/infinite-query-hook-demo.tsx

```diff
@@ -0,0 +1,147 @@
+} from '@/registry/default/blocks/infinite-query-hook/hooks/use-infinite-query'
+import { Database } from '@/registry/default/fixtures/database.types'
+import * as React from 'react'
+
+interface InfiniteListProps<TableName extends SupabaseTableName> {
+  tableName: TableName
+  columns?: string
+  pageSize?: number
+  trailingQuery?: SupabaseQueryHandler<TableName>
+  renderItem: (item: SupabaseTableData<TableName>, index: number) => React.ReactNode
+  className?: string
+  renderNoResults?: () => React.ReactNode
+  renderEndMessage?: () => React.ReactNode
+  renderSkeleton?: (count: number) => React.ReactNode
+}
+
+const DefaultNoResults = () => (
+  <div className="text-center text-muted-foreground py-10">No results.</div>
+)
+
+const DefaultEndMessage = () => (
+  <div className="text-center text-muted-foreground py-4 text-sm">You&apos;ve reached the end.</div>
+)
+
+const defaultSkeleton = (count: number) => (
+  <div className="flex flex-col gap-2 px-4">
+    {Array.from({ length: count }).map((_, index) => (
+      <div key={index} className="h-4 w-full bg-muted animate-pulse" />
+    ))}
+  </div>
+)
+
+export function InfiniteList<TableName extends SupabaseTableName>({
+  tableName,
+  columns = '*',
+  pageSize = 20,
+  trailingQuery,
+  renderItem,
+  className,
+  renderNoResults = DefaultNoResults,
+  renderEndMessage = DefaultEndMessage,
+  renderSkeleton = defaultSkeleton,
+}: InfiniteListProps<TableName>) {
+  const { data, isFetching, hasMore, fetchNextPage, isSuccess } = useInfiniteQuery({
+    tableName,
+    columns,
+    pageSize,
+    trailingQuery,
+  })
+
+  // Ref for the scrolling container
+  const scrollContainerRef = React.useRef<HTMLDivElement>(null)
+
+  // Intersection observer logic - target the last rendered *item* or a dedicated sentinel
+  const loadMoreSentinelRef = React.useRef<HTMLDivElement>(null)
+  const observer = React.useRef<IntersectionObserver | null>(null)
+
+  React.useEffect(() => {
+    if (observer.current) observer.current.disconnect()
+
+    observer.current = new IntersectionObserver(
+      (entries) => {
+        if (entries[0].isIntersecting && hasMore && !isFetching) {
+          fetchNextPage()
+        }
+      },
+      {
+        root: scrollContainerRef.current, // Use the scroll container for scroll detection
+        threshold: 0.1, // Trigger when 10% of the target is visible
+        rootMargin: '0px 0px 100px 0px', // Trigger loading a bit before reaching the end
+      }
+    )
+
+    if (loadMoreSentinelRef.current) {
+      observer.current.observe(loadMoreSentinelRef.current)
+    }
+
+    return () => {
+      if (observer.current) observer.current.disconnect()
+    }
+  }, [isFetching, hasMore, fetchNextPage])
+
+  return (
+    <div ref={scrollContainerRef} className={cn('relative h-full overflow-auto', className)}>
+      <div>
+        {isSuccess && data.length === 0 && renderNoResults()}
+
+        {data.map((item, index) => renderItem(item, index))}
```

Comment:
maybe define a rendering function instead? 

I didn't and got this trip-up
![screenshot-2025-04-14-at-16 54 52](https://github.com/user-attachments/assets/496f4d26-2c75-4a31-bf8c-442d5e54694a)


![screenshot-2025-04-14-at-16 54 11](https://github.com/user-attachments/assets/63e98606-b12c-455f-8b19-ebde37cf4368)

Reply:
This example is a an abstraction over the hook, to show you use it. I've expanded the example on how to use the abstraction.